### DISCUSSION TITLE
How to setup with blueprints?

### DISCUSSION THREAD

**Discussion by peterstavrou** [pallets-eco/flask-security#611] [Q&A] [answered]
Can someone please advise how to set this up with blueprints?
I'm having trouble adding:
```
user_datastore = SQLAlchemyUserDatastore(db, User, Role)
security = Security(app, user_datastore)
```
I'm assuming it should go in **__init__.py** and at the moment I have:
```
from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from flask_migrate import Migrate
from flask_security import Security, SQLAlchemyUserDatastore
from app.models import User, Role 
from app.config import Config

db = SQLAlchemy()
migrate = Migrate()

def create_app(config_class=Config): 
    app = Flask(__name__)
    app.config.from_object(Config) 
    app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
    app.debug = True
    db.init_app(app)
    migrate.init_app(app, db)

    user_datastore = SQLAlchemyUserDatastore(db, User, Role)
    Security.init_app(app, user_datastore)

    from app.routes import main
    app.register_blueprint(main)

    return app

#from app import models
```

I keep getting a circular import error when importing models at the top which is why I originally had it down the bottom (commented out) but I need to import User and Role from models for _SQLAlchemyUserDatastore(db, User, Role)_

`ImportError: cannot import name 'db' from partially initialized module 'app' (most likely due to a circular import)`

**Comment by jwag956** [comment, marked as answer]
Very common problem when first working with Flask!

Look at examples/unified_signin/server for one way to do this. Let me know if that helps.

**Reply by peterstavrou** [reply]
Thank you! I also really appreciate you bringing Flask-Security back to life.

For any other beginner who has this issue here is what i did...

1. Created an extensions.py file
```
from flask_security import Security
from flask_sqlalchemy import SQLAlchemy

db = SQLAlchemy()
security = Security()
```

2. Added `user_datastore = SQLAlchemyUserDatastore(db, User, Role) ` to the bottom of models.py

3. Imported them to init.py

```
from .extensions import db, security, migrate
from .models import user_datastore

def create_app(config_class=Config): 
    app = Flask(__name__)
    app.config.from_object(Config) 
    db.init_app(app)
    security.init_app(app, user_datastore)
    return app
```
